### ISSUE TITLE
Assuming a blank string for unset environment variables is problematic.

### ISSUE THREAD

**Issue by codezomb** [docker/compose#3768]
TL;DR - an empty string is not the same thing as undefined, and shouldn't be treated as such.

``` yaml

---
version: '2'
services:
  app:
    image: myrailsapplication
    build: .
    environment:
      - RAILS_ENV=${RAILS_ENV}
```

Considering the above docker-compose.yml file, If RAILS_ENV is not defined, it should not be set to an empty string. This actually causes complications within the application (development, for example), as the running code will check for an unset value, and apply certain development related settings. However, in other environments it's desirable to explicitly set this.

Sure, I could use an docker-compose.override.yml, but that's seems quite overkill. For example having to specify and maintain multiple files for a single variable setting.

**Comment by aanand** [comment]
I get what you're saying, but the current behaviour is equivalent to passing the same string to `docker run`, and should therefore be consistent with that.

```
$ docker run -e RAILS_ENV=${RAILS_ENV} alpine env | grep RAILS_ENV
RAILS_ENV=
```

Indeed, Bash works much the same:

```
$ RAILS_ENV=${RAILS_ENV} env | grep RAILS_ENV
RAILS_ENV=
```

I agree that it'd be useful to be able to differentiate between unset variables and empty values, and I'd like to see proposals for syntax/semantics, but whatever we do should remain consistent with the Docker CLI.

**Comment by T3rm1** [comment]
I want to add another example where it is required to have the variable unset instead of blank:
```
services:
  myservice:
    image: myimage
    group_add:
      - ${MY_GROUP}
```
If `MY_GROUP` is not set then it defaults to empty string and the container will not run. It will only run if the variable has a valid group id or is not set at all.
